libs/langchain/langchain/agents/agent_toolkits/conversational_retrieval/openai_functions.py

```diff
@@ -0,0 +1,80 @@
+from typing import List, Optional
+
+from langchain.agents.agent import AgentExecutor
+from langchain.agents.openai_functions_agent.agent_token_buffer_memory import (
+    AgentTokenBufferMemory,
+)
+from langchain.agents.openai_functions_agent.base import OpenAIFunctionsAgent
+from langchain.chat_models.openai import ChatOpenAI
+from langchain.memory.token_buffer import ConversationTokenBufferMemory
+from langchain.prompts.chat import MessagesPlaceholder
+from langchain.schema.language_model import BaseLanguageModel
+from langchain.schema.memory import BaseMemory
+from langchain.schema.messages import SystemMessage
+from langchain.tools.base import BaseTool
+
+
+def _get_default_system_message() -> SystemMessage:
+    return SystemMessage(
+        content=(
+            "Do your best to answer the questions. "
+            "Feel free to use any tools available to look up "
+            "relevant information, only if neccessary"
+        )
+    )
+
+
+def create_conversational_retrieval_agent(
+    llm: BaseLanguageModel,
+    tools: List[BaseTool],
+    remember_intermediate_steps: bool = True,
+    memory_key: str = "chat_history",
+    system_message: Optional[SystemMessage] = None,
+    verbose: bool = False,
+) -> AgentExecutor:
+    """A convenience method for creating a conversational retrieval agent.
+
+    Args:
+        llm: The language model to use, should usually be ChatOpenAI
+        tools: A list of tools the agent has access to
+        remember_intermediate_steps: Whether the agent should remember intermediate
+            steps or not. Intermediate steps refer to prior action/observation
+            pairs from previous questions. The benefit of remembering these is if
+            there is relevant information in there, the agent can use it to answer
+            follow up questions. The downside is it will take up more tokens.
+        memory_key: The name of the memory key in the prompt.
+        system_message: The system message to use. By default, a basic one will
+            be used.
+        verbose: Whether or not the final AgentExecutor should be verbose or not,
+            defaults to False.
+
+    Returns:
+        An agent executor initialized appropriately
+    """
+
+    if not isinstance(llm, ChatOpenAI):
+        raise ValueError("Only supported with ChatOpenAI models.")
+    if remember_intermediate_steps:
```

Comment:
imo should be able to optionally pass in memory

```python
if memory:
  pass
elif remember_intermediate_steps:
  memory = AgentToken...
else:
  memory = ConversationTokenBuff...

Reply:
eh this is just a constructor, and the notebook spells out pretty easily how to construct your own